core/src/client/client.rs

```diff
@@ -694,17 +697,17 @@ impl ReopenBlock for Client {
 }
 
 impl PrepareOpenBlock for Client {
-    fn prepare_open_block(&self, author: Address, extra_data: Bytes) -> OpenBlock {
+    fn prepare_open_block(&self, parent_block: BlockId, author: Address, extra_data: Bytes) -> OpenBlock {
         let engine = &*self.engine;
         let chain = self.block_chain();
-        let h = engine.get_block_hash_to_mine_on(chain.best_block_hash());
-        let latest_header = &chain.block_header(&h).expect("h is best block hash: so its header must exist: qed");
+        let h = self.block_hash(&parent_block).expect("parent exist always");
+        let parent_header = chain.block_header(&h).expect("parent exist always");
```

Comment:
```rust
assert_eq!(h, parent_header.hash());
```

Reply:
Do we need the assert?
`chain.block_header` has the responsibility to satisfy the assertion.

Reply:
I think adding `debug_assert_eq` and using `h` at line 706 is better.